xla/service/gpu/gpu_windowed_einsum_handler_test.cc

```diff
@@ -193,5 +197,146 @@ ENTRY main.9_spmd {
+  param.4 = bf16[2,512,24576]{2,1,0} parameter(0), sharding={devices=[1,4,1]<=[4]}
+  param.5 = bf16[24576,24576]{1,0} parameter(1), sharding={devices=[1,4]<=[4]}
+  constant.22 = bf16[] constant(0)
+  broadcast = bf16[2,2048,24576]{2,1,0} broadcast(constant.22), dimensions={}
+  constant.24 = u32[] constant(0)
+  tuple.2 = (bf16[2,512,24576]{2,1,0}, bf16[24576,24576]{1,0}, bf16[2,2048,24576]{2,1,0}, bf16[2,2048,24576]{2,1,0}, u32[]) tuple(param.4, param.5, broadcast, broadcast, constant.24)
+  while = (bf16[2,512,24576]{2,1,0}, bf16[24576,24576]{1,0}, bf16[2,2048,24576]{2,1,0}, bf16[2,2048,24576]{2,1,0}, u32[]) while(tuple.2), condition=windowed_dot_general_cond_ag, body=windowed_dot_general_body_ag
+  get-tuple-element.13 = bf16[2,2048,24576]{2,1,0} get-tuple-element(while), index=2
+  copy.1 = bf16[2,2048,24576]{2,1,0} copy(get-tuple-element.13)
+  all-gather = bf16[2,2048,24576]{2,1,0} all-gather(param.4), channel_id=1, replica_groups={{0,1,2,3}}, dimensions={1}, use_global_device_ids=true
+  param.6 = bf16[24576,24576]{1,0} parameter(2), sharding={devices=[1,4]<=[4]}
+  dot.7 = bf16[2,2048,24576]{2,1,0} dot(all-gather, param.6), lhs_contracting_dims={2}, rhs_contracting_dims={0}
+  add.13 = bf16[2,2048,24576]{2,1,0} add(copy.1, dot.7)
+  param.7 = bf16[24576,24576]{1,0} parameter(3), sharding={devices=[4,1]<=[4]}
+  broadcast.2 = bf16[2,512,24576]{2,1,0} broadcast(constant.22), dimensions={}
+  tuple.3 = (bf16[2,2048,24576]{2,1,0}, bf16[24576,24576]{1,0}, bf16[2,512,24576]{2,1,0}, bf16[2,512,24576]{2,1,0}, u32[]) tuple(add.13, param.7, broadcast.2, broadcast.2, constant.24)
+  while.1 = (bf16[2,2048,24576]{2,1,0}, bf16[24576,24576]{1,0}, bf16[2,512,24576]{2,1,0}, bf16[2,512,24576]{2,1,0}, u32[]) while(tuple.3), condition=windowed_dot_general_cond_rs, body=windowed_dot_general_body_rs
+  get-tuple-element.14 = bf16[2,512,24576]{2,1,0} get-tuple-element(while.1), index=2
+  collective-permute.4 = bf16[2,512,24576]{2,1,0} collective-permute(get-tuple-element.14), channel_id=7, source_target_pairs={{0,1},{1,2},{2,3},{3,0}}
+  get-tuple-element.15 = bf16[2,512,24576]{2,1,0} get-tuple-element(while.1), index=3
+  ROOT add.14 = bf16[2,512,24576]{2,1,0} add(collective-permute.4, get-tuple-element.15)
+}
+)";
+
+  TF_ASSERT_OK_AND_ASSIGN(std::unique_ptr<HloModule> module,
+                          ParseAndReturnVerifiedModule(kHloString));
+
+  GpuWindowedEinsumHandler gpu_handler;
+  bool changed;
+  TF_ASSERT_OK_AND_ASSIGN(changed, gpu_handler.Run(module.get()));
+  EXPECT_TRUE(changed);
+
+  HloInstruction* ag_loop =
+      FindInstructionByName(module->entry_computation(), "while");
+  HloInstruction* inst =
+      FindInstructionByName(module->entry_computation(), "dot.7");
+  // dot.7 should now consume output of the windowed einsum while loop.
+  EXPECT_TRUE(inst->operand(0)->opcode() == HloOpcode::kGetTupleElement);
+  EXPECT_TRUE(inst->operand(0)->tuple_index() == 3);
+  EXPECT_TRUE(inst->operand(0)->operand(0) == ag_loop);
```

Comment:
Is there an execution test?

Reply:
updated

Reply:
CollectiveOpsTestE2E.WindowedEinsumE2EAllgatherMultiConsumer under //xla/tests:collective_ops_test_e2e